notebooks/network-analysis.ipynb

```diff
@@ -0,0 +1,434 @@
+    "subreachable = nhd_network.reachable(rconn)"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "## Visualize subnetworks\n",
+    "- Make a map showing each of the subnetworks in the supernetwork with different colors. "
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "# Get the tailwater keys for each subnetwork \n",
+    "n_subnets = subreachable.keys()\n",
+    "\n",
+    "# initialize a new column in the GeoDataFrame that will containa subnetwork code\n",
+    "gdf[\"subnetwork\"] = 0\n",
+    "\n",
+    "# loop through subnetworks, populate the newly created \"subnetwork\" columm with an integer subnetwork identifier code\n",
+    "i = 1\n",
+    "for sub in n_subnets:\n",
+    "    \n",
+    "    # get the stream segments in subnetwork `sub`\n",
+    "    A =  [abs(ele) for ele in list(subreachable[sub])]\n",
+    "    \n",
+    "    # find the subnetwork rows in GeoDataFrame \n",
+    "    C = gdf.index.isin(A)\n",
+    "    \n",
+    "    # Insert a subnetwork code in the `subnetwork` column of the GeoDataFrame\n",
+    "    gdf.at[C,\"subnetwork\"] = i\n",
+    "    \n",
+    "#     print(i, sum(C), sub)\n",
+    "    \n",
+    "    i = i+1\n",
+    "    \n",
+    "# Map creation ...\n",
+    "# ----------------------------------------------\n",
+    "\n",
+    "# CONUS basemap\n",
+    "base = world[world.name == \"United States of America\"].plot(\n",
+    "    color='white', edgecolor='black')\n",
+    "\n",
+    "# add our stream network, using different marker colors for each subnetwork\n",
+    "gdf.plot(ax = base, markersize = 0.01, column = \"subnetwork\", cmap = 'prism')\n",
+    "\n",
+    "# crop the map extent\n",
+    "base.set_xlim(-130,-60)\n",
+    "base.set_ylim(20,55)\n",
+    "\n",
+    "plt.title('The supernetwork broken into subnetworks')\n",
+    "plt.show()"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "## Identify reaches\n",
+    "- Identifying reaches in each subnetwork is a two step process\n",
+    "    1. Use the `reachable_network` function to identify the upstream connectivity structure of each subnetwork. The function accepts the reverse network dictionary, `rconn`, and returns a dictionary, `subnets_`, which lists upstream connections for each segment in each subnetwork. \n",
+    "    2. Conduct a [depth-first-search](https://en.wikipedia.org/wiki/Depth-first_search) of each subnetwork. This is done with the `dfs_decomposition` function, which accepts the forward (downstream oriented) network dictionary, `nhd_network.reverse_network(net)`, and returns a dictionary of segment lists defining specific reaches in the subnetwork. Results from each subnetwork are concatenated into a larger dictionary, `subreaches`, which contains segment lists for all reaches in the supernetwork.    \n"
```

Comment:
Big note here: dfs_decomposition doesn't reverse the network anymore as of yesterday. It operates on the network you pass in.